### PR TITLE
Make π < π work as expected

### PR DESCRIPTION
**Pull request by perrutquist**
Methods are added for < and <= comparing Irrationals. (The corresponding
methods for > and >= are then handled by existing fallback methods.)

Currently, expressions like `ℯ < π` result in promotion to Float64.
This works most of the time. However, `π < π` does not result in
promotion and throws an error. This commit introduces methods
for comparing an Irrational to itself.

This commit also introduces a test to make sure that incorrect results
are not being returned due to different user-defined irrationals
rounding to the same Float64.

### CODE REVIEW

**Review by simonbyrne — APPROVED** [review]

File: base/irrationals.jl [hunk 1]
```diff
@@ -56,6 +56,17 @@ float(::Type{<:AbstractIrrational}) = Float64
 ==(::Irrational{s}, ::Irrational{s}) where {s} = true
 ==(::AbstractIrrational, ::AbstractIrrational) = false
 
+<(::Irrational{s}, ::Irrational{s}) where {s} = false
+function <(x::AbstractIrrational, y::AbstractIrrational)
+    # If two different Irrationals round to the same Float64, then explicit
+    # < methods are needed for those inputs, as the below assertation will fail.
+    @assert Float64(x) != Float64(y)
```
Comment (martinholters) [inline base/irrationals.jl]: This should probably throw some specific error.

**Comment by simonbyrne** [comment]
I thought it did, but I just tried various versions back to 0.4 and apparently not!

File: base/irrationals.jl (hunk 1 above)
Comment (simonbyrne) [reply base/irrationals.jl]: Probably a good idea. Something like
```
Float64(x) != Float64(y) || error("$x and $y have identical Float64 representations")
```

Comment (martinholters) [reply base/irrationals.jl]: Or a `MethodError`. The solution is to add a specific method, after all.

Comment (perrutquist) [reply base/irrationals.jl]: I've added a commit, changing the code to throw a `MethodError`.

**Comment by waldyrious** [comment]
Cross-referencing the Discourse thread where this issue was raised: https://discourse.julialang.org/t/comparison-of-same-irrationals-by-throws-error/1342/

**Pull request merged**